File: cytnx/contraction.cpp
#include "cytnx.hpp"

using namespace cytnx;

int main(){
    int dim = 1000;
    std::vector<std::vector<cytnx_int64>> qns;
    for(int i = 0; i < dim; i++){
        qns.push_back({i});
    }   
    Bond b1 = Bond(dim, BD_KET, qns);
    Bond b2 = Bond(3, BD_KET, {{1},{2},{3}});
    Bond b3 = Bond(3, BD_BRA, {{1},{2},{3}});
    UniTensor T1 = UniTensor({b1, b2, b3}, {0, 1, 2}, 1);
    UniTensor T2 = UniTensor({b2, b3}, {2, 3}, 1);
    Contract(T1, T2);

    return 0;
}

File: deprecated/cytnx/contraction.py
import cytnx
import time

dims = [100, 200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000]
sparse, dense = [], []
for dim in dims:
    qns = []
    for i in range(dim):
        qns.append([1])
        qns.append([-1])
        
    b1 = cytnx.Bond(2*dim, cytnx.BD_KET, qns)
    b2 = cytnx.Bond(2*dim, cytnx.BD_KET, qns)

    T1 = cytnx.UniTensor([b1, b2.redirect()], labels = [0, 1], rowrank = 1)
    T2 = cytnx.UniTensor([b1, b2.redirect()], labels = [1, 2], rowrank = 1)

    sumt = 0
    for k in range(10):
        st = time.time()
        _ = cytnx.Contract(T1, T2)
        en = time.time()
        print("Sparse contraction time : ", en-st)
        sumt += (en-st)
    sparse.append(sumt/10)
    
    T3 = cytnx.UniTensor(cytnx.zeros([2*dim, 2*dim]), rowrank=1)
    T3 = T3.relabels([0, 1])
    T4 = cytnx.UniTensor(cytnx.zeros([2*dim, 2*dim]), rowrank=1)
    T4 = T4.relabels([1, 2])

    sumt = 0
    for k in range(10):
        st = time.time()
        _ = cytnx.Contract(T3, T4)
        en = time.time()
        print("Dense contraction time : ", en-st)
        sumt += (en-st)
    dense.append(sumt/10)

    print("="*20)

print("Sparse contraction times : ", sparse)
print("Dense contraction times : ", dense)
